File: leet-code/isValid.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function (s) {
  let stack = []
  for (let i = 0; i < s.length; i++) {
    if (s[i] === '{' || s[i] === '[' || s[i] === '(') {
      // 进栈
      stack.push(s[i])
    }
    if (s[i] === '}' || s[i] === ']' || s[i] === ')') {
      let element = stack.pop()
      if (
        (s[i] === '}' && element !== '{') ||
        (s[i] === ']' && element !== '[') ||
        (s[i] === ')' && element !== '(')
      ) {
        return false
      }
    }
  }
  return stack.length > 0 ? false : true
}

console.log(isValid('('))

File: leet-code/isValid.go
package main

func isValid(s string) bool {
	var stack []rune
	for _, val := range s {
		if val == '[' || val == '{' || val == '(' {
			stack = append(stack, val)
		}
		if val == ']' || val == '}' || val == ')' {
			// 如果进来的时候栈为空，也不对
			if len(stack) == 0 {
				return false
			}
			temp := stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			if (val == ']' && temp != '[') || (val == '}' && temp != '{') || (val == ')' && temp != '(') {
				return false
			}
		}
	}
	if len(stack) > 0 {
		return false
	}
	return true
}
